File: processor.c
#include <erl_nif.h>

static ERL_NIF_TERM
avg(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
    return enif_make_int(env, 2);
}

static ERL_NIF_TERM
max(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
    return enif_make_int(env, 3);
}

static ERL_NIF_TERM
min(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
    return enif_make_int(env, 1);
}

static ErlNifFunc nif_funcs[] =
{
    {"avg", 1, avg},
    {"max", 1, max},
    {"min", 1, min}
};

ERL_NIF_INIT(Elixir.Processor,nif_funcs,NULL,NULL,NULL,NULL)

File: processor.ex
defmodule Processor do
  @on_load :load_nifs

  def load_nifs do
    :erlang.load_nif('./processor', 1)
  end

  def avg(_) do
    raise "NIF avg/1 not implemented"
  end

  def max(_) do
    raise "NIF max/1 not implemented"
  end

  def min(_) do
    raise "NIF min/1 not implemented"
  end

  def commands, do: %{ "avg" => &Processor.avg/1, "max" => &Processor.max/1, "min" => &Processor.min/1 }
end
